url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/cmdlib/BeanLibrary.java

```diff
@@ -0,0 +1,560 @@
+        }
+    }
+
+    private static Object resolveCanonical(List<String> aPieces, int aCurrPiece, Object aBase)
+    throws CommandException
+    {
+        if(aPieces.size() <= aCurrPiece) return aBase;
+        final String lPiece = aPieces.get(aCurrPiece);
+        int lIdx = -1;
+        try
+        {
+            lIdx = Integer.parseInt(lPiece);
+        }
+        catch (NumberFormatException e1)
+        {
+            // Ignore this exception!
+        }
+
+        if(aBase instanceof IContext && lIdx < 0)
+        {
+            final IContext lCtx = (IContext) aBase;
+            if(lCtx.isBound(lPiece))
+                return resolveCanonical(aPieces, aCurrPiece + 1, lCtx.getBinding(lPiece));
+            else
+                throw new CommandException(String.format("Cannot find '%s' in context.", lPiece));
+        }
+        else if(aBase instanceof Map && lIdx < 0)
+        {
+            final Map lMap = (Map) aBase;
+            if(lMap.containsKey(lPiece))
+                return resolveCanonical(aPieces, aCurrPiece + 1, lMap.get(lPiece));
+            else
+                throw new CommandException("ERROR - map.");
+        }
+        else if(aBase instanceof List)
+        {
+            final List lList = (List) aBase;
+            if(lIdx < 0) throw new CommandException("List expects an index.");
+            if(lIdx >= lList.size()) throw new CommandException("List index out of range.");
+            else return resolveCanonical(aPieces, aCurrPiece + 1, lList.get(lIdx));
+        }
+        else  if(aBase != null)
+        {
+            if(aBase.getClass().isArray())
+            {
+                // Array rendering.
+                ///////////////////
+
+                if(lIdx >=0 && lIdx < Array.getLength(aBase))
+                    return resolveCanonical(aPieces, aCurrPiece + 1, Array.get(aBase, lIdx));
+                else throw new CommandException("ERROR - index out of range/wrong index type/no index.");
+            }
+            else
+            {
+                try
+                {
+                    // Try bean.
+                    //
+                    final BeanInfo lInfo = Introspector.getBeanInfo(aBase.getClass());
+                    final PropertyDescriptor lProps[] = lInfo.getPropertyDescriptors();
+                    for (PropertyDescriptor lProp : lProps)
+                    {
+                        if (lProp.getName().equals(lPiece))
+                        {
+                            Method lGetter = lProp.getReadMethod();
+
+                            Object[] lArgs;
+                            if(lIdx >= 0) lArgs = new Object[]{lIdx};
+                            else  lArgs = new Object[]{};
+
+                            Object lVal = lGetter.invoke(aBase, lArgs);
+                            return resolveCanonical(aPieces, aCurrPiece + 1, lVal);
+                        }
+                    }
+                }
+                catch(CommandException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)